File: mac/FluentCore/Tests/FluentCoreTests/StyleFormatterTests.swift
import Testing
@testable import FluentCore

/// The Android StyleFormatterTest cases, unchanged, so both platforms format identically.
struct StyleFormatterTests {
    func fmt(_ text: String, _ style: WritingStyle, _ category: StyleCategory = .other) -> String {
        StyleFormatter.format(text, style: style, category: category)
    }

    let lunch = "Hey, are you free for lunch tomorrow? Let's do 12 if that works for you."

    @Test func formalLeavesSmartTranscriptAlone() {
        #expect(fmt(lunch, .formal, .personal) == lunch)
    }

    @Test func casualDropsIntroCommaAndFinalPeriod() {
        #expect(fmt(lunch, .casual, .personal) == "Hey are you free for lunch tomorrow? Let's do 12 if that works for you")
    }

    @Test func veryCasualLowercasesAndStripsButKeepsQuestions() {
        #expect(fmt(lunch, .veryCasual, .personal) == "hey are you free for lunch tomorrow? let's do 12 if that works for you")
    }

    @Test func veryCasualKeepsPronounAcronymsAndInTokenPunctuation() {
        #expect(fmt("I think NASA said 3.5 at 12:30 on fluent.app, and 1,000 people came.", .veryCasual)
            == "I think NASA said 3.5 at 12:30 on fluent.app and 1,000 people came")
        #expect(fmt("Ok, see you there! Wait, what?!", .veryCasual) == "ok see you there wait what?")
    }

    @Test func excitedOnlyChangesClosingFullStop() {
        #expect(fmt("So far, I am enjoying it. I am excited for tomorrow.", .excited)
            == "So far, I am enjoying it. I am excited for tomorrow!")
        #expect(fmt("Are you coming?", .excited) == "Are you coming?")
        #expect(fmt("No punctuation here", .excited) == "No punctuation here")
    }

    @Test func abbreviationsAndEllipsesKeepTheirDots() {
        #expect(fmt("See you at 5 p.m.", .casual) == "See you at 5 p.m.")
        #expect(fmt("See you at 5 p.m.", .excited) == "See you at 5 p.m.")
        #expect(fmt("Well...", .casual) == "Well...")
    }

    @Test func casualKeepsNumbersAndNamedGreetings() {
        #expect(fmt("Hi Alex, it costs 1,000 dollars.", .casual, .work) == "Hi Alex, it costs 1,000 dollars")
    }

    let email = "Hi Alex, it was great talking with you today. Looking forward to our next chat. Best, Mary."

    @Test func formalEmailIsLaidOut() {
        #expect(fmt(email, .formal, .email)
            == "Hi Alex,\n\nIt was great talking with you today. Looking forward to our next chat.\n\nBest,\nMary")
    }

    @Test func excitedEmailExclaimsBodyNotSignOff() {
        #expect(fmt(email, .excited, .email)
            == "Hi Alex,\n\nIt was great talking with you today. Looking forward to our next chat!\n\nBest,\nMary")
    }

    @Test func casualEmailKeepsGreetingInline() {
        #expect(fmt(email, .casual, .email)
            == "Hi Alex, it was great talking with you today. Looking forward to our next chat\n\nBest, Mary")
    }

    @Test func emailLayoutOnlyInEmailApps() {
        #expect(fmt(email, .formal, .other) == email)
    }

    @Test func emailWithoutGreetingOrSignOffIsNotRestructured() {
        #expect(fmt("Quarterly report attached.", .formal, .email) == "Quarterly report attached.")
        #expect(StyleFormatter.parseEmail("Hey I was wondering about the report.") == nil)
    }

    @Test func emailGreetingNeverSwallowsPronounI() {
        #expect(fmt("Hi, I wanted to follow up on Friday. Kind regards, Sam Lee.", .formal, .email)
            == "Hi,\n\nI wanted to follow up on Friday.\n\nKind regards,\nSam Lee")
    }

    @Test func existingLineBreaksSurvive() {
        #expect(fmt("First line.\n\nSecond line.", .veryCasual) == "first line\n\nsecond line")
    }

    @Test func blankInputStaysBlank() {
        #expect(fmt("   ", .casual) == "")
    }

    @Test func everyCategoryDefaultIsOneOfItsStyles() {
        for c in StyleCategory.allCases { #expect(c.styles.contains(c.defaultStyle), "\(c)") }
    }

    // Cases the Android suite does not have: non-ASCII text and curly apostrophes.
    @Test func unicodeTextSurvivesVeryCasual() {
        #expect(fmt("Café, it’s great. Ünter den Linden!", .veryCasual) == "café it’s great ünter den Linden")
        #expect(fmt("I’m here.", .veryCasual) == "I’m here")
    }
}

File: tests/Fluent.Core.Tests/StyleFormatterTests.cs
using Fluent.Core;
using Xunit;

namespace Fluent.Core.Tests;

/// <summary>The Android StyleFormatterTest cases (via the Mac suite), unchanged, so every platform
/// formats identically.</summary>
public class StyleFormatterTests
{
    static string Fmt(string text, WritingStyle style, StyleCategory category = StyleCategory.Other) =>
        StyleFormatter.Format(text, style, category);

    const string Lunch = "Hey, are you free for lunch tomorrow? Let's do 12 if that works for you.";

    [Fact] public void FormalLeavesSmartTranscriptAlone() =>
        Assert.Equal(Lunch, Fmt(Lunch, WritingStyle.Formal, StyleCategory.Personal));

    [Fact] public void CasualDropsIntroCommaAndFinalPeriod() =>
        Assert.Equal("Hey are you free for lunch tomorrow? Let's do 12 if that works for you",
            Fmt(Lunch, WritingStyle.Casual, StyleCategory.Personal));

    [Fact] public void VeryCasualLowercasesAndStripsButKeepsQuestions() =>
        Assert.Equal("hey are you free for lunch tomorrow? let's do 12 if that works for you",
            Fmt(Lunch, WritingStyle.VeryCasual, StyleCategory.Personal));

    [Fact]
    public void VeryCasualKeepsPronounAcronymsAndInTokenPunctuation()
    {
        Assert.Equal("I think NASA said 3.5 at 12:30 on fluent.app and 1,000 people came",
            Fmt("I think NASA said 3.5 at 12:30 on fluent.app, and 1,000 people came.", WritingStyle.VeryCasual));
        Assert.Equal("ok see you there wait what?", Fmt("Ok, see you there! Wait, what?!", WritingStyle.VeryCasual));
    }

    [Fact]
    public void ExcitedOnlyChangesClosingFullStop()
    {
        Assert.Equal("So far, I am enjoying it. I am excited for tomorrow!",
            Fmt("So far, I am enjoying it. I am excited for tomorrow.", WritingStyle.Excited));
        Assert.Equal("Are you coming?", Fmt("Are you coming?", WritingStyle.Excited));
        Assert.Equal("No punctuation here", Fmt("No punctuation here", WritingStyle.Excited));
    }

    [Fact]
    public void AbbreviationsAndEllipsesKeepTheirDots()
    {
        Assert.Equal("See you at 5 p.m.", Fmt("See you at 5 p.m.", WritingStyle.Casual));
        Assert.Equal("See you at 5 p.m.", Fmt("See you at 5 p.m.", WritingStyle.Excited));
        Assert.Equal("Well...", Fmt("Well...", WritingStyle.Casual));
    }

    [Fact] public void CasualKeepsNumbersAndNamedGreetings() =>
        Assert.Equal("Hi Alex, it costs 1,000 dollars", Fmt("Hi Alex, it costs 1,000 dollars.", WritingStyle.Casual, StyleCategory.Work));

    const string Email = "Hi Alex, it was great talking with you today. Looking forward to our next chat. Best, Mary.";

    [Fact] public void FormalEmailIsLaidOut() =>
        Assert.Equal("Hi Alex,\n\nIt was great talking with you today. Looking forward to our next chat.\n\nBest,\nMary",
            Fmt(Email, WritingStyle.Formal, StyleCategory.Email));

    [Fact] public void ExcitedEmailExclaimsBodyNotSignOff() =>
        Assert.Equal("Hi Alex,\n\nIt was great talking with you today. Looking forward to our next chat!\n\nBest,\nMary",
            Fmt(Email, WritingStyle.Excited, StyleCategory.Email));

    [Fact] public void CasualEmailKeepsGreetingInline() =>
        Assert.Equal("Hi Alex, it was great talking with you today. Looking forward to our next chat\n\nBest, Mary",
            Fmt(Email, WritingStyle.Casual, StyleCategory.Email));

    [Fact] public void EmailLayoutOnlyInEmailApps() => Assert.Equal(Email, Fmt(Email, WritingStyle.Formal, StyleCategory.Other));

    [Fact]
    public void EmailWithoutGreetingOrSignOffIsNotRestructured()
    {
        Assert.Equal("Quarterly report attached.", Fmt("Quarterly report attached.", WritingStyle.Formal, StyleCategory.Email));
        Assert.Null(StyleFormatter.ParseEmail("Hey I was wondering about the report."));
    }

    [Fact] public void EmailGreetingNeverSwallowsPronounI() =>
        Assert.Equal("Hi,\n\nI wanted to follow up on Friday.\n\nKind regards,\nSam Lee",
            Fmt("Hi, I wanted to follow up on Friday. Kind regards, Sam Lee.", WritingStyle.Formal, StyleCategory.Email));

    [Fact] public void ExistingLineBreaksSurvive() =>
        Assert.Equal("first line\n\nsecond line", Fmt("First line.\n\nSecond line.", WritingStyle.VeryCasual));

    [Fact] public void BlankInputStaysBlank() => Assert.Equal("", Fmt("   ", WritingStyle.Casual));

    [Fact]
    public void EveryCategoryDefaultIsOneOfItsStyles()
    {
        foreach (var c in Enum.GetValues<StyleCategory>()) Assert.Contains(c.DefaultStyle(), c.Styles());
    }

    [Fact]
    public void UnicodeTextSurvivesVeryCasual()
    {
        Assert.Equal("café it’s great ünter den Linden", Fmt("Café, it’s great. Ünter den Linden!", WritingStyle.VeryCasual));
        Assert.Equal("I’m here", Fmt("I’m here.", WritingStyle.VeryCasual));
    }

    // Windows-only cases: transcripts arrive with Windows line endings from the clipboard path.
    [Fact] public void SamplesFormatForEveryStyle()
    {
        foreach (var c in Enum.GetValues<StyleCategory>())
            foreach (var s in c.Styles())
                Assert.False(string.IsNullOrWhiteSpace(StyleFormatter.Format(c.Sample(), s, c)));
    }
}
